Modules/_sqlite/connection.c

```diff
@@ -1866,6 +1866,53 @@ pysqlite_connection_exit_impl(pysqlite_Connection *self, PyObject *exc_type,
     Py_RETURN_FALSE;
 }
 
+/*[clinic input]
+_sqlite3.Connection.setlimit as setlimit
+
+    limit: int
+        The limit category to be set.
+    value: int
+        The new limit. If the new limit is a negative number, the limit is
+        unchanged.
+    /
+
+Set connection run-time limits. Non-standard.
+
+Attempts to increase a limit above its hard upper bound are silently truncated
+to the hard upper bound. Regardless of whether or not the limit was changed,
+the prior value of the limit is returned.
+[clinic start generated code]*/
+
+static PyObject *
+setlimit_impl(pysqlite_Connection *self, int limit, int value)
+/*[clinic end generated code: output=18d15e4be8a7a4ec input=e0990e0c90e747ee]*/
+{
+    if (!pysqlite_check_thread(self) || !pysqlite_check_connection(self)) {
+        return NULL;
+    }
+
+    int previous_value = sqlite3_limit(self->db, limit, value);
+    return PyLong_FromLong(previous_value);
```

Comment:
I am wondering what `sqlite3_limit()` returns for invalid limit category and for closed sqlite3?

Reply:
It returns -1 for invalid limit category. If `self->db == NULL`, it segfaults (unless you've compiled SQLite with `SQLITE_ENABLE_API_ARMOR`; then it returns -1). If the connection is closed, a random value is sometimes returned. The `sqlite3` module always clears `self->db` upon `sqlite3_close_v2`, and `pysqlite_check_connection()` does a sanity check on `self->db`, so we know we're always calling `sqlite3_limit` with a valid connection pointer.